(cvt, dshl) Exception in thread "main"

### Checklist

- [x] Did you specify the current behavior?
- [x] Did you specify the expected behavior?
- [x] Did you provide a code example showing the problem?
- [x] Did you describe your environment?
- [ ] Did you specify relevant external information?

### What is the current behavior?

The following FIRRTL program.

```python
; seed: 7897
circuit top_mod :
  module top_mod :
    input inp_3: SInt<1>
    output tmp17: SInt<13>
    wire tmp9: SInt<1>
    tmp9 <= cvt(inp_3)
    tmp17 <= dshl(SInt<12>(-822), asUInt(tmp9))
```

Running with `firrtl-1.4.0` crashes with the following error:

```
Exception in thread "main" firrtl.FirrtlInternalException: Internal Error! Please file an issue at https://github.com/ucb-bar/firrtl/issues
        at firrtl.Utils$.error(Utils.scala:454)
        at firrtl.Utils$.throwInternalError(Utils.scala:163)
        at firrtl.stage.phases.CatchExceptions.transform(CatchExceptions.scala:38)
        at firrtl.stage.phases.CatchExceptions.transform(CatchExceptions.scala:17)
        at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:38)
        at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:15)
        at firrtl.options.Translator.transform(Phase.scala:248)
        at firrtl.options.Translator.transform$(Phase.scala:248)
        at firrtl.options.phases.DeletedWrapper.transform(DeletedWrapper.scala:15)
        at firrtl.options.DependencyManager.$anonfun$transform$3(DependencyManager.scala:278)
        at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:168)
        at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:164)
        at scala.collection.immutable.List.foldLeft(List.scala:79)
        at firrtl.options.DependencyManager.transform(DependencyManager.scala:269)
        at firrtl.options.DependencyManager.transform$(DependencyManager.scala:255)
        at firrtl.options.PhaseManager.transform(DependencyManager.scala:436)
        at firrtl.stage.FirrtlStage.run(FirrtlStage.scala:38)
        at firrtl.options.Stage$$anon$1.transform(Stage.scala:43)
        at firrtl.options.Stage$$anon$1.transform(Stage.scala:43)
        at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:38)
        at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:15)
        at firrtl.options.Translator.transform(Phase.scala:248)
        at firrtl.options.Translator.transform$(Phase.scala:248)
        at firrtl.options.phases.DeletedWrapper.transform(DeletedWrapper.scala:15)
        at firrtl.options.Stage.$anonfun$transform$5(Stage.scala:47)
        at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:168)
        at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:164)
        at scala.collection.immutable.List.foldLeft(List.scala:79)
        at firrtl.options.Stage.$anonfun$transform$3(Stage.scala:47)
        at logger.Logger$.$anonfun$makeScope$2(Logger.scala:166)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at logger.Logger$.makeScope(Logger.scala:164)
        at firrtl.options.Stage.transform(Stage.scala:47)
        at firrtl.options.Stage.execute(Stage.scala:58)
        at firrtl.options.StageMain.main(Stage.scala:71)
        at firrtl.stage.FirrtlMain.main(FirrtlStage.scala)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at coursier.bootstrap.launcher.a.a(Unknown Source)
        at coursier.bootstrap.launcher.Launcher.main(Unknown Source)
Caused by: java.util.NoSuchElementException
        at scala.collection.immutable.BitmapIndexedMapNode.apply(HashMap.scala:617)
        at scala.collection.immutable.HashMap.apply(HashMap.scala:132)
        at firrtl.VerilogEmitter$.firrtl$VerilogEmitter$$precedenceGt(Emitter.scala:273)
        at firrtl.VerilogEmitter.parenthesize$1(Emitter.scala:465)
        at firrtl.VerilogEmitter.op_stream(Emitter.scala:550)
        at firrtl.VerilogEmitter.emitCol(Emitter.scala:373)
        at firrtl.VerilogEmitter.$anonfun$emitCol$2(Emitter.scala:412)
        at firrtl.VerilogEmitter.$anonfun$emitCol$2$adapted(Emitter.scala:411)
        at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:168)
        at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:164)
        at scala.collection.immutable.List.foldLeft(List.scala:79)
        at firrtl.VerilogEmitter.emitCol(Emitter.scala:411)
        at firrtl.VerilogEmitter.$anonfun$emitCol$2(Emitter.scala:412)
        at firrtl.VerilogEmitter.$anonfun$emitCol$2$adapted(Emitter.scala:411)
        at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:168)
        at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:164)
        at scala.collection.immutable.List.foldLeft(List.scala:79)
        at firrtl.VerilogEmitter.emitCol(Emitter.scala:411)
        at firrtl.VerilogEmitter.emit(Emitter.scala:339)
        at firrtl.VerilogEmitter$VerilogRender.$anonfun$emit_streams$8(Emitter.scala:1317)
        at firrtl.VerilogEmitter$VerilogRender.$anonfun$emit_streams$8$adapted(Emitter.scala:1317)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:920)
        at firrtl.VerilogEmitter$VerilogRender.emit_streams(Emitter.scala:1317)
        at firrtl.VerilogEmitter$VerilogRender.emit_verilog(Emitter.scala:1435)
        at firrtl.VerilogEmitter.$anonfun$emit$2(Emitter.scala:1476)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at firrtl.VerilogEmitter.emit(Emitter.scala:1470)
        at firrtl.VerilogEmitter.$anonfun$execute$4(Emitter.scala:1488)
        at scala.collection.immutable.List.flatMap(List.scala:293)
        at scala.collection.immutable.List.flatMap(List.scala:79)
        at firrtl.VerilogEmitter.execute(Emitter.scala:1485)
        at firrtl.Transform.transform(Compiler.scala:312)
        at firrtl.Transform.transform$(Compiler.scala:312)
        at firrtl.SeqTransform.transform(Compiler.scala:394)
        at firrtl.stage.transforms.ExpandPrepares.execute(ExpandPrepares.scala:19)
        at firrtl.Transform.transform(Compiler.scala:312)
        at firrtl.Transform.transform$(Compiler.scala:312)
        at firrtl.stage.transforms.ExpandPrepares.transform(ExpandPrepares.scala:7)
        at firrtl.stage.transforms.CatchCustomTransformExceptions.execute(CatchCustomTransformExceptions.scala:10)
        at firrtl.Transform.transform(Compiler.scala:312)
        at firrtl.Transform.transform$(Compiler.scala:312)
        at firrtl.stage.transforms.CatchCustomTransformExceptions.transform(CatchCustomTransformExceptions.scala:7)
        at firrtl.stage.transforms.UpdateAnnotations.$anonfun$internalTransform$1(UpdateAnnotations.scala:22)
        at firrtl.Utils$.time(Utils.scala:169)
        at firrtl.Transform$.runTransform(Compiler.scala:210)
        at firrtl.stage.transforms.UpdateAnnotations.internalTransform(UpdateAnnotations.scala:22)
        at firrtl.stage.transforms.UpdateAnnotations.internalTransform(UpdateAnnotations.scala:8)
        at firrtl.options.Translator.transform(Phase.scala:248)
        at firrtl.options.Translator.transform$(Phase.scala:248)
        at firrtl.stage.transforms.UpdateAnnotations.transform(UpdateAnnotations.scala:8)
        at firrtl.options.DependencyManager.$anonfun$transform$3(DependencyManager.scala:278)
        at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:168)
        at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:164)
        at scala.collection.immutable.List.foldLeft(List.scala:79)
        at firrtl.options.DependencyManager.transform(DependencyManager.scala:269)
        at firrtl.options.DependencyManager.transform$(DependencyManager.scala:255)
        at firrtl.stage.TransformManager.transform(TransformManager.scala:14)
        at firrtl.stage.phases.Compiler.$anonfun$internalTransform$6(Compiler.scala:120)
        at firrtl.Utils$.time(Utils.scala:169)
        at firrtl.stage.phases.Compiler.f$1(Compiler.scala:120)
        at firrtl.stage.phases.Compiler.$anonfun$internalTransform$8(Compiler.scala:125)
        at scala.collection.immutable.List.map(List.scala:246)
        at scala.collection.immutable.List.map(List.scala:79)
        at firrtl.stage.phases.Compiler.internalTransform(Compiler.scala:125)
        at firrtl.stage.phases.Compiler.internalTransform(Compiler.scala:45)
        at firrtl.options.Translator.transform(Phase.scala:248)
        at firrtl.options.Translator.transform$(Phase.scala:248)
        at firrtl.stage.phases.Compiler.transform(Compiler.scala:45)
        at firrtl.stage.phases.CatchExceptions.transform(CatchExceptions.scala:30)
        ... 39 more
```

### What is the expected behavior?

Not to crash. Give a more informative Error or Warning or just work.

### Steps to Reproduce

`./firrtl-1.4.0 -i top_mod.fir`

### Your environment

`Linux linux-sydi 5.9.14-1-default #1 SMP Sat Dec 12 06:57:32 UTC 2020 (c648a46) x86_64 x86_64 x86_64 GNU/Linux`

- Verilator version: `Verilator 4.103 devel rev v4.102-42-g726e78fd`

### External Information

<!-- Was this discussed anywhere else, e.g., Twitter, Gitter, StackOverflow? Provide direct links if available! -->

--- a/src/main/scala/firrtl/backends/verilog/VerilogEmitter.scala
+++ b/src/main/scala/firrtl/backends/verilog/VerilogEmitter.scala
@@ -57,6 +57,13 @@ object VerilogEmitter {
   private def precedenceGt(op1: PrimOp, op2: PrimOp): Boolean = {
     precedenceMap(op1) < precedenceMap(op2)
   }
+
+  /** Identifies PrimOps that never need parentheses
+    *
+    * These PrimOps emit either {..., a0, ...} or a0 so they never need parentheses
+    */
+  private val neverParens: PrimOp => Boolean =
+    Set(Shl, Cat, Cvt, AsUInt, AsSInt, AsClock, AsAsyncReset, Pad)
 }
 
 class VerilogEmitter extends SeqTransform with Emitter {
@@ -229,8 +236,7 @@ class VerilogEmitter extends SeqTransform with Emitter {
   // to ensure Verilog operations are signed.
   def op_stream(doprim: DoPrim): Seq[Any] = {
     def parenthesize(e: Expression, isFirst: Boolean): Any = doprim.op match {
-      // these PrimOps emit either {..., a0, ...} or a0 so they never need parentheses
-      case Shl | Cat | Cvt | AsUInt | AsSInt | AsClock | AsAsyncReset => e
+      case op if neverParens(op) => e
       case _ =>
         e match {
           case e: DoPrim =>
@@ -247,7 +253,8 @@ class VerilogEmitter extends SeqTransform with Emitter {
                */
               case other =>
                 val noParens =
-                  precedenceGt(e.op, doprim.op) ||
+                  neverParens(e.op) ||
+                    precedenceGt(e.op, doprim.op) ||
                     (isFirst && precedenceEq(e.op, doprim.op) && !isUnaryOp(e.op))
                 if (noParens) other else Seq("(", other, ")")
             }

--- a/src/test/scala/firrtlTests/InlineBooleanExpressionsSpec.scala
+++ b/src/test/scala/firrtlTests/InlineBooleanExpressionsSpec.scala
@@ -392,6 +392,22 @@ class InlineBooleanExpressionsSpec extends FirrtlFlatSpec {
     firrtlEquivalenceTest(input, Seq(new InlineBooleanExpressions))
   }
 
+  // https://github.com/chipsalliance/firrtl/issues/2035
+  // This is interesting because other ways of trying to express this get split out by
+  // SplitExpressions and don't get inlined again
+  // If we were to inline more expressions (ie. not just boolean ones) the issue this represents
+  // would come up more often
+  it should "handle cvt nested inside of a dshl" in {
+    val input =
+      """circuit DshlCvt:
+        |  module DshlCvt:
+        |    input a: UInt<4>
+        |    input b: SInt<1>
+        |    output o: UInt
+        |    o <= dshl(a, asUInt(cvt(b)))""".stripMargin
+    firrtlEquivalenceTest(input, Seq(new InlineBooleanExpressions))
+  }
+
   it should s"respect --${PrettyNoExprInlining.longOption}" in {
     val input =
       """circuit Top :

